### PR TITLE
Create system indices if absent

### PR DESCRIPTION
**Pull request by wangch079**
System indices (currently `.elastic-connectors` and `.elastic-connectors-sync-jobs`) are created at Kibana startup. But connector service can't assume the existence of Kibana, and needs to create those indices if they are absent.

#### Pre-Review Checklist
- [x] Covered the changes with automated tests
- [x] Tested the changes locally

### CODE REVIEW

File: lib/app/connector.rb [hunk 1]
```diff
@@ -28,6 +28,8 @@ def start!
 
       def pre_flight_check
         raise "#{App::Config['service_type']} is not a supported connector" unless Connectors::REGISTRY.registered?(App::Config['service_type'])
+        Framework::ElasticConnectorActions.ensure_connectors_index_exists
```
Comment (wangch079) [inline lib/app/connector.rb:31]: This won't stop the app from exiting as there won't be any entry in the connector index.

File: lib/framework/elastic_connector_actions.rb [hunk 2]
```diff
@@ -12,88 +12,137 @@
...
+    JOB_INDEX = '.elastic-connectors-sync-jobs'
 
-    def self.force_sync(connector_package_id)
-      body = {
-        :doc => {
-          :scheduling => { :enabled => true },
-          :sync_now => true
+    class << self
+
+      def force_sync(connector_package_id)
```
Comment (artem-shelkovnikov) [inline lib/framework/elastic_connector_actions.rb:19]: While I agree with refactoring (`class << self`), I think we should avoid having multiple various changes in a single PR - it should be split to a different PR and reviewed separately.

I see it as a recurring pattern - when PR is small and starts to expand, adding multiple various changes which slows down the review/merge procedure + increase the changes/size of merge conflicts.

File: lib/framework/elastic_connector_actions.rb [hunk 3]
```diff
@@ -12,88 +12,137 @@
...
-    end
+      # should only be used in CLI
+      def ensure_index_exists(index_name, body = {})
+        client.indices.create(:index => index_name, :body => body) unless client.indices.exists?(:index => index_name)
+      end
 
-    # should only be used in CLI
-    def self.ensure_connectors_index_exists
-      ensure_index_exists(CONNECTORS_INDEX)
+      def ensure_alias_exists(alias_name, body = {})
```
Comment (artem-shelkovnikov) [inline lib/framework/elastic_connector_actions.rb:97]: What's up with the alias? Why do we have/need it?

**Comment by artem-shelkovnikov** [comment]
@wangch079 can you provide a PR description too, as to what problem is being solved?

File: lib/framework/elastic_connector_actions.rb [hunk 4]
```diff
@@ -12,88 +12,137 @@
...
+        @client ||= Utility::EsClient.new
+      end
 
-    # should only be used in CLI
-    def self.ensure_index_exists(index_name)
-      client.indices.create(:index => index_name) unless client.indices.exists?(:index => index_name)
-    end
+      # should only be used in CLI
+      def ensure_index_exists(index_name, body = {})
+        client.indices.create(:index => index_name, :body => body) unless client.indices.exists?(:index => index_name)
```
Comment (seanstory) [inline lib/framework/elastic_connector_actions.rb:94]: Should this check to make sure the response is a 20x? Otherwise `ensure_index_exists` isn't really ensuring anything, it's just attempting to create it.

**Review by seanstory — COMMENTED** [review]
As a side note, we should be careful about our verbiage. These indices are not "system indices" - that's a special term for Elasticsearch-managed indices with special permissions. Right now, I think they're just normal indices. But they could be hidden indices.

File: lib/framework/elastic_connector_actions.rb (hunk 2 above)
Comment (wangch079) [reply lib/framework/elastic_connector_actions.rb:19]: Agree. Will avoid doing so.

File: lib/framework/elastic_connector_actions.rb (hunk 3 above)
Comment (wangch079) [reply lib/framework/elastic_connector_actions.rb:97]: For migration purposes. See [here](https://docs.google.com/document/d/1_4VCIDhj-a3EVHIJ_tADm28AtqZdZvo4CW-vrYAWcIQ/edit#bookmark=id.iqzlyufomwsk)

File: lib/framework/elastic_connector_actions.rb (hunk 4 above)
Comment (wangch079) [reply lib/framework/elastic_connector_actions.rb:94]: If it's not 20x, I think it will just raise an error, and the app/CLI will exit. We should probably catch it and display more user-friendly message. cc @artem-shelkovnikov @mchernyavskaya

File: lib/framework/elastic_connector_actions.rb [hunk 5]
```diff
@@ -12,88 +12,137 @@
...
-      ensure_index_exists(CONNECTORS_INDEX)
+      def ensure_alias_exists(alias_name, body = {})
+        body[:aliases] = { alias_name => { :is_write_index => true } }
+        client.indices.create(:index => "#{alias_name}_v1", :body => body) unless client.indices.exists?(:index => alias_name)
+      end
+
+      # should only be used in CLI
+      def ensure_connectors_index_exists
+        body = {
+          :mappings => {
```
[the remark this answers is not in the document]
Comment (wangch079) [reply lib/framework/elastic_connector_actions.rb:105]: Good point

File: lib/framework/elastic_connector_actions.rb (hunk 3 above)
Comment (artem-shelkovnikov) [reply lib/framework/elastic_connector_actions.rb:97]: TIL, thanks for the explanation! In this case we might need to add logic to find out which version index we need to create based on what's the last known version, but probably could be done later

Comment (wangch079) [reply lib/framework/elastic_connector_actions.rb:97]: If the alias/index does not exist, we can assume we need to create `_v1` index. 

I would like to keep this PR simple, as there can be a lot of edge cases, e.g. alias not existing but index existing. I added it as one of the topic to discuss in [Engineers sync](https://docs.google.com/document/d/1Z3t79RpKpML-4vsuFAavXzMqAcw_cEbDqoK7Ov33hVg/edit#)

File: lib/framework/elastic_connector_actions.rb (hunk 4 above)
Comment (wangch079) [reply lib/framework/elastic_connector_actions.rb:94]: @artem-shelkovnikov, the connector service doesn't handle edge cases well. I think it's fine as we want to move fast. We can have a dedicated PR later to make the connector service more robust.

Comment (artem-shelkovnikov) [reply lib/framework/elastic_connector_actions.rb:94]: Good call, 100% agree

**Review by artem-shelkovnikov — APPROVED** [review]

File: lib/framework/elastic_connector_actions.rb (hunk 4 above)
Comment (seanstory) [reply lib/framework/elastic_connector_actions.rb:94]: For when you get to that client augmentation, I suggest looking at [this code](https://github.com/elastic/ent-search/blob/efeb730e1dda78631b7b8a47ed8ecc35c917f239/lib/swiftype/es/client.rb#L14-L30). I'm currently leveraging this Elasticsearch Client/wrapper for the Elasticsearch Crawler, and really liking how it works. I think it might be a good candidate for us to open so that we don't have to reinvent it from scratch.
